### PR TITLE
C++/C#: Alias analysis of indirect parameters

### PR DESCRIPTION
**Pull request by dbartol**
This change enables more precise alias analysis the memory pointed to by pointer and reference parameters. At a high level, we now treat the memory pointed to by the value of a pointer/reference parameter to be its own unaliased allocation, independent of all other allocations. This is unsound, but produces much better analysis results in practice.

Overall, IR diffs (in aliased SSA) are what I would expect. Stores to an indirect parameter no longer pollute the all aliased memory virtual variable. Even though we don't know how big any given indirect parameter buffer is, if all accesses to that buffer are just scalars at offset zero, we'll still hook up uses to exact definitions just as if we knew the size all along.

The memory accessed by the `InitializeIndirection` and `ReturnIndirection` instructions now has a new `MemoryAccessKind`: `EntireAllocationMemoryAccess`. This kind represents an access to the entire contiguous allocation, without knowing exactly how big that allocation is.

Alias analysis now operates on `Allocation`s, which are provided as module arguments to represent the memory locations to be tracked. In unaliased SSA, an `Allocation` is just an `IRAutomaticVariable`. In aliased SSA, there is an `Allocation` for each `IRVariable`, plus one for each indirect parameter.

Aliased SSA still works mostly the same, except that a new type of `MemoryLocation` has been added to represent an access to the entire `Allocation`. The overlap computation has been updated accordingly.

The existing `AliasAnalysis.qll` was shared between IR stages in C++, but was not automatically shared with C#. I brought the two languages in sync, so that now they will be kept up to date as they change in the future. This required adding a couple more predicates to the language interface. C# did not already have the same "models" system as C++, so for now, I implemented a trivial `AliasFunction` class that never actually has any object instances.

I've implemented a new module for writing analysis tests called `InlineExpectationTests.qll`. You extend `InlineExpectationsTest` to specify what results you found on each line. Each result has a tag and a value, both strings. The tags and values are automatically compared to comments in the code under test to determine which results are expected, which are unexpected, and which are missing. The output file only shows missing and unexpected results, so a passing test should have an empty `.expected` file. There's also a way to annotate a particular result as being either a false positive or false negative. This results in a different message if the actual result doesn't match the expected. For example, if the actual output is missing a result that was declared as a false positive, the output message reports that a false positive was fixed.

The new `points_to.ql` test uses the above module.

### CODE REVIEW

**Comment by dbartol** [comment]
I've rebased onto `master`, and curated the PR into several commits. The commits aren't expected to work correctly on their own, but they should at least separate all the changes into related chunks for easier review.

File: cpp/ql/test/TestUtilities/InlineExpectationsTest.qll [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+      message = "Invalid expectation syntax: " + expectation.getExpectation()
+    )
+  }
+}
+
+private string expectationCommentPattern() { result = "//\\s*(\\$(?:[^/]|/[^/])*)(?://.*)?" }
+
+private string expectationPattern() {
+  result = "(?:(f(?:\\+|-)):)?((?:[A-Za-z-_]+)(?:\\s*,\\s*[A-Za-z-_]+)*)(?:=(.*))?"
+}
```
[the remark this answers is not in the document]
Comment (dbartol) [reply cpp/ql/test/TestUtilities/InlineExpectationsTest.qll]: I've documented each of them individually (with examples in the module header as part of the "how to use this" docs).

File: cpp/ql/test/TestUtilities/InlineExpectationsTest.qll [hunk 2]
```diff
@@ -0,0 +1,168 @@
+import cpp
+
+abstract class InlineExpectationsTest extends string {
```
[the remark this answers is not in the document]
Comment (dbartol) [reply cpp/ql/test/TestUtilities/InlineExpectationsTest.qll]: I've added a huge module comment with details and examples.

**Comment by dbartol** [comment]
@rdmarsh2 I think I've addressed all of your feedback.

**Comment by rdmarsh2** [comment]
sync-identical-files failed because #2713 touched the C++ copies of AliasAnalysis.qll
